interface/ceed-basis.c

```diff
@@ -1249,20 +1249,23 @@ int CeedQRFactorization(Ceed ceed, CeedScalar *mat, CeedScalar *tau,
       sigma += v[j] * v[j];
     }
     CeedScalar norm = sqrt(v[i]*v[i] + sigma); // norm of v[i:m]
-    CeedScalar Rii = -copysign(norm, v[i]);
-    v[i] -= Rii;
+    CeedScalar R_ii = -copysign(norm, v[i]);
+    v[i] -= R_ii;
+    mat[i+n*i] = R_ii;
     // norm of v[i:m] after modification above and scaling below
     //   norm = sqrt(v[i]*v[i] + sigma) / v[i];
     //   tau = 2 / (norm*norm)
     tau[i] = 2 * v[i]*v[i] / (v[i]*v[i] + sigma);
 
+    // Don't apply Householder reflector if row already diagonal
+    if (sigma < CEED_EPSILON) continue;
```

Comment:
I think you mean that the column is already diagonal, but I'm still not convinced this is good or desirable. Perhaps it's harmless for our unblocked algorithm, but it's unusual and breaks the mathematical purity of using computed factors, including the semantics of blocking. (Sure, we may not plan to implement blocked/CompactWY methods.) My concern is not so much that it's wrong as written for the matrices we're working with, but that it's no better and thus spurious complexity that someone down the road will grapple with and after painstaking testing, rip out.

From the definition of `sigma` up above, I think this condition can trigger if you have a well-behaved non-triangular matrix that has been multiplied by 1e-15. I.e., to work with general matrices, any such tolerance must be suitably scaled relative to the norm of the column (acknowledging that `sigma` has been squared). And I still don't think this logic should be needed.

Reply:
If I just remove it, 7 tests fail with
```
/home/jeremy/Dev/libCEED/interface/ceed-basis.c:1422 in CeedSymmetricSchurDecomposition(): Symmetric QR failed to converge
```

Reply:
How does it factor `[2e-10   -1e-10; -1e-10  2e-10]`?

Reply:
It makes a disaster
```
7.071068e-01	5.000000e-01	5.000000e-01	0.000000e+00	
-7.071068e-01	5.000000e-01	5.000000e-01	0.000000e+00	
0.000000e+00	-5.000000e-01	5.000000e-01	-7.071068e-01	
0.000000e+00	-5.000000e-01	5.000000e-01	7.071068e-01
```

Reply:
Currently, what we have in main is broken, as it takes in diagonal matrices and un-diagonalizes them.

Reply:
I'm lost. Shouldn't the result be 2x2?

Reply:
I'm trying to re-acquire my context from April. I don't remember the full series of changes and observations that lead me to this change.

I don't think (might be wrong though) this sigma check I added matters for the QR factorization, but it seems to matter for using the same algorithm from the QR factorization for the tridiagonalization inside of the symmetric Schur decomposition. But I'm not seeing why that  should make a difference; as far as I can tell, if it works ok for the QR then it should work ok as part of the symmetric Schur. So I figure that I must be missing something about this version of QR. Unfortunately, your modifications make it different enough from my copy of Golub and Van Loan that whatever is going wrong is not obvious to me.

But the problem in the tridiagonalization only shows up with a 4x4 matrix or larger, so I ran the 4x4 version of the test case you proposed. The test case did emit a correct decomposition that when multiplied together did in fact recreate the original test case matrix.

Reply:
The example I gave is catastrophic if you run the old code (with the condition to skip the reflection). First I made the test print accurately so we can see.
<details>

```diff
diff --git i/tests/t301-basis.c w/tests/t301-basis.c
index b8de71ee..3d96152b 100644
--- i/tests/t301-basis.c
+++ w/tests/t301-basis.c
@@ -5,19 +5,17 @@
 
 int main(int argc, char **argv) {
   Ceed ceed;
-  CeedScalar qr[12] = {1, -1, 4, 1, 4, -2, 1, 4, 2, 1, -1, 0};
-  CeedScalar tau[3];
+  CeedScalar qr[4] = {2e-10, -1e-10, -1e-10, 2e-10};
+  CeedScalar tau[2];
 
   CeedInit(argv[1], &ceed);
 
-  CeedQRFactorization(ceed, qr, tau, 4, 3);
-  for (int i=0; i<12; i++) {
-    if (qr[i] <= 1E-14 && qr[i] >= -1E-14) qr[i] = 0;
-    fprintf(stdout, "%12.8f\n", qr[i]);
+  CeedQRFactorization(ceed, qr, tau, 2, 2);
+  for (int i=0; i<4; i++) {
+    fprintf(stdout, "%12.8e\n", qr[i]);
   }
-  for (int i=0; i<3; i++) {
-    if (tau[i] <= 1E-14 && tau[i] >= -1E-14) tau[i] = 0;
-    fprintf(stdout, "%12.8f\n", tau[i]);
+  for (int i=0; i<2; i++) {
+    fprintf(stdout, "%12.8e\n", tau[i]);
   }
   CeedDestroy(&ceed);
   return 0;
```
</details>

then (with the conditional early continue)
```console
$ build/t301-basis /cpu/self
-2.23606798e-10
-1.00000000e-10
-1.00000000e-10
-2.00000000e-10
1.89442719e+00
2.00000000e+00
```
and without
```console
$ build/t301-basis /cpu/self
-2.23606798e-10
1.78885438e-10
-2.36067977e-01
-1.34164079e-10
1.89442719e+00
2.00000000e+00
```
These are totally different and the first one doesn't represent an orthogonal matrix because `I - tau v v^T` is not a reflector when `tau=1.894427189` and `v=[1, -1e-10]`.

I don't understand why this fails for the diagonalization, but a correct QR should not early continue like this.

Reply:
BTW, LAPACK skips the _last_ reflection (via `tau=0`) for square matrices because there's nothing below the diagonal.
```
julia> using LinearAlgebra
julia> A = [2 -1; -1 2] * 1e-10; tau = [0.,0]; LAPACK.geqrf!(A, tau)
([-2.23606797749979e-10 1.7888543819998312e-10; -0.23606797749978967 1.341640786499874e-10], [1.8944271909999157, 0.0])
```